go-controller/pkg/node/gateway_localnet.go

```diff
@@ -239,89 +247,377 @@ func localnetIPTablesHelper(subnet *net.IPNet) (util.IPTablesHelper, error) {
 	return ipt, nil
 }
 
-func localnetIptRules(svc *kapi.Service, gatewayIP string) []iptRule {
-	rules := make([]iptRule, 0)
-	for _, svcPort := range svc.Spec.Ports {
-		protocol, err := util.ValidateProtocol(svcPort.Protocol)
+type activeSocket interface {
+	Close() error
+}
+
+type localNodePortWatcher interface {
+	initIPTables() error
+	getLocalAddrs() error
+	addService(svc *kapi.Service) error
+	deleteService(svc *kapi.Service) error
+	syncServices(serviceInterface []interface{})
+}
+
+type localnetNodePortWatcherData struct {
+	localNodePortWatcher
+	recorder      record.EventRecorder
+	ipt           util.IPTablesHelper
+	gatewayIP     string
+	localAddrSet  map[string]net.IPNet
+	activeSockets map[int32]activeSocket
+}
+
+func newLocalnetNodePortWatcherData(ipt util.IPTablesHelper, gatewayIP net.IP, recorder record.EventRecorder) *localnetNodePortWatcherData {
+	return &localnetNodePortWatcherData{
+		ipt:           ipt,
+		gatewayIP:     gatewayIP.String(),
+		recorder:      recorder,
+		activeSockets: make(map[int32]activeSocket),
+	}
+}
+
+func (npw *localnetNodePortWatcherData) openLocalPort(port int32, protocol kapi.Protocol, svc *kapi.Service) error {
+	var socket activeSocket
+	var socketError error
+	switch protocol {
+	case kapi.ProtocolTCP:
```

Comment:
Probably should add a TODO for SCTP support.

Reply:
Indeed, I will update now

Reply:
Yes, it's on L302 as `unsupported` :smiley: 

I've been looking at SCTP go libraries, and the only one I could find is: https://github.com/ishidawataru/sctp 

But I would prefer we leave it `unsupported` for know and have another PR address the SCTP socket creation

Reply:
*Something something this PR is growing out of proportion*

Reply:
Ohh I see what could have caused confusion, I added the `unsupported` code to the commit `Adding tests for physical and virtual node port setup` which is why you might not have seen it in the preceding commit. Will try to change that

Reply:
@squeed @alexanderConstantinescu the more I think about this the more I think this is not a good idea. We don't need to bind ports for OVN-k8s to work, traffic will hit netfilter first and get SNAT'ed to OVN before it ever hits a userspace bound process. So this opening of N number of ports is simply in place as a guard for other applications that may exist on the host. I think this is kind of unnecessary, because:
1) anyone launching applications on the host, should be aware that ports are in use by kubernetes, since this node is being used as a k8s host
2) if someone creates a host network pod and defines a port, then we can check if that port is in use by any existing service and issue a warning or fail. I wonder if k8s already checks if the port range used by a host network pod is already in the node port range  (30000 - 32767) and warns of conflict?

I don't think the solution is for us to bind to ports, but rather catch the issue when a pod is created that may conflict. Otherwise the admin of the host is free to do whatever he wants at his own risk.

Reply:
@girishmg @dcbw @danwinship would like to hear your thoughts on ^ as well

Reply:
> I wonder if k8s already checks if the port range used by a host network pod is already in the node port range  (30000 - 32767) and warns of conflict?

K8s does what this PR adds: https://github.com/kubernetes/kubernetes/blob/master/pkg/proxy/iptables/proxier.go#L1039

One thing I however came to realize is that if we decide to merge this, I will need to change how this is done. As it stands now it will only be done in local gateway mode.

Reply:
Yes, k8s upstream behaves this way, and it has historically been considered important, to keep people from being able to snipe ports from each other and from system services. Sure, we could specifically check that hostports don't conflict with service ports, but there could be other undeclared ports in use, or ports that are declared as part of some new resource type we don't know about. Making everyone fall back to claiming an actual kernel listening socket makes everything consistent...

Reply:
OK if that is the k8s upstream way, then we should follow that. I still don't agree with it, but let's be consistent. @alexanderConstantinescu please also add support for shared gateway mode.

Reply:
Will do, @trozet